README
	Manitou-Mail v1.0.0 - a database-driven mail user agent

    Copyright(C) 2004-2010 Daniel Verite <daniel@manitou-mail.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 
   as published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   For more details see the file COPYING.

What's Manitou-Mail
===================

Manitou is a database-driven mail user-agent and associated utilities.
It provides a framework and tools for storing and exposing mail as
organized data, accessible by SQL. It is aimed at handling seamlessly
large mailboxes, possibly processed by multiple concurrent users.  In
addition to the features commonly found in traditional mail programs,
its user interface comes with some more unique, database oriented ways
to organize, process and retrieve mail messages.

Visit the project's website: http://www.manitou-mail.org

Build notes
===========
In order to build from sources pulled from the version control repository,
the configure script and Makefile.in files need to be generated by the
autotools.

The steps to remake them are:

$ aclocal		# creates aclocal.m4
$ autoheader		# creates config.h.in
$ autoconf		# creates configure
$ automake -a -c	# creates various Makefile.in and scripts

The project requires automake v1.6 at least. If your default installed
version is less recent, you may need to look for `aclocal-1.6' and
`automake-1.6' instead of `aclocal' and `automake'.


NEWS
 v1.0.2, Sep 23 2010
--------------------------
user interface
- bugfix in tags with hierarchies set by filter rules
- automatic database reconnection
- headers are no longer printed on a separate page
- enable clipboard copy in the attached mail view.

manitou-mdx
- new config parameter log_filter_hits

 v1.0.1, Aug 20 2010
--------------------------
user interface
- instant notification of new mail
- window title now displays the database name
- dynamic signatures
- bugfixes on message counters
- fixed possible crash when mime_types was empty

manitou-mgr
- new fields on users table
- populate the mime_types table at initialization

 v1.0.0, Aug 11 2010
--------------------------
documentation
- improved and converted from html to docbook format

user interface
- full-featured HTML composer based on webkit
- global notepad
- improved main listview

manitou-mdx
- multipart/alternative supported for outgoing mail
- more Unix-like standard daemon behavior (fork, pidfile, syslog)
- automatic reconnect to the database and other improvements in robustness
- notify for new mail
- new plugin spam_learner that automates calls to sa-learn

 v0.9.12, Jul 9 2009
--------------------------
user interface
- new email address completion and headers layout in the email composer
- bugfixes on empty attachments
- bugfixes in html viewer
- partial load for large messages

manitou-mgr
- new commands to create database, schemas and query the inverted word index

manitou-mdx
- new plugin: trashcan_purge
- fixed missing full text indexing for outgoing messages

 v0.9.11, Dec 6 2008
--------------------------
user interface
- use webkit to display HTML and text contents
- show filter logs

manitou-mdx and database
- support for content_id in attachments
- enhancements in filters: logs, new actions and functions

 v0.9.10, Nov 4 2008
--------------------------
user interface
- full Qt4 port, no longer depends on the Qt3support library

manitou-mdx and database
- lots of referential integrity constraints
- merged trashed_mail and mail tables
- tnef_decoder and pdfindexer new plugins
- new print-size option for manitou-mgr

 v0.9.9, Dec 9 2007
--------------------------
user interface
- easier selection of a temporary identity when composing mail
- shared attachment contents for outgoing mails, based on SHA1 fingeprint

manitou-mdx
- new manitou-mgr utility with hash-attachments and merge-attachments actions
- test for the existence of an attachment within the database before importing
(shared attachments)

 v0.9.8, Nov 18 2007
--------------------------
user interface
- Qt4 port
- improvements on attachment handling
- various bugfixes
- Windows version builds on MinGW

manitou-mdx
- improved mixing large imports with other processing
- accept multiple encodings for outgoing mail
- compute SHA1 fingerprint of attachments

 v0.9.7, Apr 22 2006
--------------------------
user interface
- bugfix on preferred font loading in composition window
- builds on SunOS and *BSD systems and with gcc-2.95.x
- zoom on sql sentence in query selection dialog
- progress bar for attachment downloads
- improved and enriched help system using Qt assistant
- improved URLs recognition in message body
- new "Fetching" section in Preferences
- optional pre-fetching of messages bodies when idle
- improved tag editor with drag and drop support
- allow the display of raw headers when available
- bugfix in preferences on the attachments download path
- bugfix in filters editor when the focus is still in an
  edited field on validation.
- revamped, faster search engine

manitou-mdx
- new config parameter 'detach_text_plain' to transfer unnamed
text/plain attached parts into the message body
- speed up the bayesian classifier plugin by caching statistics
- optionally save raw mail files into large objects
- improved parsing of non-conformant Date fields.
- modularization
- new build scripts conforming to standard Perl practice

 v0.9.6, May 20 2005
---------------------
manitou-mdx
- admin plugins with built-in scheduler
- don't barf on unknown encoding in headers
- new bayesian classifier plugin
- new attachments uploader outgoing plugin
- fix charset of outgoing mail for unicode databases
- fix use of default temp directory
- avoid newlines in mail.subject field
- take timezones into account for sender dates

user interface
- french translation.
- web URLs are optionally clickable in messages bodies.
- tag counters are updated real-time in the quick query panel.
- tags are sorted in the quick query panel.
- when composing under Windows, timezone format has been fixed for
  RFC compliance.
- query execution times are displayed in the status bar.
- bugfix in selection of tagged messages in contextual menu.
- fallback to first identity when no default is defined in message
  composition.
- space bar now scrolls a page when the focus is inside the body panel.
- text encapsulating forwarded messages is now user-configurable.
- spurious newline after the signature is removed.
- bugfix on wrapping a line on tabs at display time.
- added default identity selection in preferences dialog.
- keep attachments directory in composition window.
- message bodies can be edited and saved.
- properly disconnect non-primary db connections.
- improved status refresh in threaded display.
- improved text extraction in "reply (quote selected text)" command.
- optional auto-selection of recipient when forwarding.
- improved quoting of messages containing several text/plain attachments.

 v0.9.5, March 24 2005
-----------------------
manitou-mdx
- accept attachment filenames containing unencoded 8 bits characters.
- new option --skip for mbox partial import.
- correct use of index_words parameter for outgoing mail.
- new plugin (mswordindexer) using vwware (http://wvware.sourceforge.net)
  to index msword attachments.
- bugfix for busy loop when the no_send parameter is used.

user interface
- windows version available.

main.h
/* Copyright (C) 2004-2010 Daniel Verite

   This file is part of Manitou-Mail (see http://www.manitou-mail.org)

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2 as
   published by the Free Software Foundation.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330,
   Boston, MA 02111-1307, USA.
*/

#ifndef INC_MAIN_H
#define INC_MAIN_H

#define TODO_WEBKIT

#ifdef _MSC_VER
#pragma warning (disable:4786)
#define strcasecmp stricmp
#define strncasecmp strnicmp
#endif

#include "config.h"
#include "dbtypes.h"

#ifndef CHECK_PTR
#define CHECK_PTR(p) do { if (!p) abort(); } while(0)
#endif

#include <QString>
#include <QApplication>
#include <QDebug>

class QStringList;
class mail_header;

#include <map>
//#include <string>

#if defined(__GNUG__) //&& !defined(QT_WS_WIN)
  #define _UNUSED_ __attribute((unused))
  #if __GNUG__>=3
  // standard (gcc3) way for defining a variable number of arguments
  #define DBG_PRINTF(l,...) debug_printf(l,__FILE__,__LINE__, __VA_ARGS__)
  #define ERR_PRINTF(...) err_printf(__FILE__,__LINE__, __VA_ARGS__)
  #else
  // or gcc2's way
  #define DBG_PRINTF(l,args...) debug_printf(l,__FILE__,__LINE__, ## args)
  #define ERR_PRINTF(args...) err_printf(__FILE__,__LINE__, ## args)
  #endif
#else
  // not gcc
  #define DBG_PRINTF {do { } while(0);}
  #define ERR_PRINTF {do { } while(0);}
  #define _UNUSED_
#endif

void debug_printf(int level, const char* file, int line, const char *fmt, ...);
void err_printf(const char* file, int line, const char *fmt, ...);
//void debug_printf1(int level, const char* file, int line, const char *fmt, ...);

#define APP_NAME "Manitou"

class FT
{
public:
  enum searchInEnum {
    searchInBodies=1,
    searchInSubjects=2,
//    searchInAuthors=4,
    searchInHeaders=4,
    searchMax=4
  };
  enum searchOptionsEnum {
    caseInsensitive=1,
    wrapAround=2,
    autoClose=4,
    optionsMax=4
  };
};

class manitou_application : public QApplication
{
public:
  manitou_application(int& argc, char** argv);
  ~manitou_application();
  void start_new_mail(const mail_header&);
  void set_style(const QString style);
  QString m_dbname;
private:
  QString default_style_name;
};

extern manitou_application* gl_pApplication;

namespace service_f
{
    QString toCodingDb(const QString &s);
    QString GetNameDb (const char* conninfo);
}

#endif // INC_MAIN_H
